dali/pipeline/operators/argument.h

```diff
@@ -31,13 +32,14 @@ class Value {
  public:
   virtual std::string ToString() const = 0;
   template <typename T>
-  static inline Value * construct(const T& val);
+  static inline std::shared_ptr<Value> construct(const T& val);
```

Comment:
I know this is legacy code, but this function looks a bit like devilish construct...
If I understand correctly, we have a virtual class, which can have one and only one subclass, despite it's construction routine (yet, the subclass is not marked as `final`, but that's not the case here). Maybe its good opportunity to repair it? I'm not an expert here, but I think, that `Value` and `ValueInst` can be merged into one class. @mzient, what do you think?

Reply:
I would leave it for refactoring. I want to address memory leaks here.

Reply:
I think that Value should be returned as unique_ptr - If the owner of that pointer decides to share it, that's his choice, but the semantics for construct would be clearer that way.

Reply:
Done

Reply:
@szalpal: It's bad, but not as bad as you described. Value is subclassed by all template instantiations and there are many of them (or: should be, because for some unknown reason we create Value<int64_t> for all integral and enum types).

Reply:
Unknown reason being Python does not understand proper enums and will always give int there, so we need to be able to retrieve all enums from arguments that are stored as ints.

Reply:
This code is basically implementation of std::any (we don't have c++17 so needed to write it myself).